qiskit/providers/aer/openpulse/hamiltonian_model.py

```diff
@@ -0,0 +1,313 @@
+
+
+class HamiltonianModel():
+    """Hamiltonian model for pulse simulator."""
+    def __init__(self, hamiltonian, qubits=None):
+        """Initialize a Hamiltonian model.
+
+        Args:
+            hamiltonian (dict): Hamiltonian dictionary.
+            qubits (list or None): List of qubits to extract from the hamiltonian.
+
+        Raises:
+            ValueError: if arguments are invalid.
+        """
+        # Initialize internal variables
+        # The system Hamiltonian in numerical format
+        self._system = None
+        # System variables
+        self._vars = None
+        # Channels in the Hamiltonian string
+        # These tell the order in which the channels are evaluated in
+        # the RHS solver.
+        self._channels = None
+        # Diagonal elements of the hamiltonian
+        self._h_diag = None
+        # Eigenvalues of the time-independent hamiltonian
+        self._evals = None
+        # Eigenstates of the time-indepedent hamiltonian
+        self._estates = None
+        # Qubit subspace dimensinos
+        self._dim_qub = {}
+        # Oscillator subspace dimensions
+        self._dim_osc = {}
+
+        # Parse Hamiltonian
+        # TODO: determine n_qubits from hamiltonian if qubits is None
+        n_qubits = len(qubits) if qubits else None
+        if not n_qubits:
+            raise ValueError("TODO: Need to infer n_qubits from "
+                             "Hamiltonian if qubits list is not specified")
+
+        self._vars = OrderedDict(hamiltonian['vars'])
+
+        # Get qubit subspace dimensions
+        if 'qub' in hamiltonian:
+            self._dim_qub = {
+                int(key): val
+                for key, val in hamiltonian['qub'].items()
+            }
+        else:
+            self._dim_qub = {}.fromkeys(range(n_qubits), 2)
+
+        # Get oscillator subspace dimensions
+        if 'osc' in hamiltonian:
+            self._dim_osc = {
+                int(key): val
+                for key, val in hamiltonian['osc'].items()
+            }
+
+        # Step 1: Parse the Hamiltonian
+        system = HamiltonianParser(h_str=hamiltonian['h_str'],
+                                   dim_osc=self._dim_osc,
+                                   dim_qub=self._dim_qub)
+        system.parse(qubits)
+        self._system = system.compiled
+
+        # Step #2: Determine Hamiltonian channels
+        self._get_hamiltonian_channels()
+
+        # Step 3: Calculate diagonal hamiltonian
+        self._get_diag_hamiltonian()
+
+    def initial_state(self):
+        """Return the initial state of the time-independent hamiltonian"""
+        # Set initial state
+        ground_state = 0 * op.basis(len(self._evals), 1)
+        for idx, estate_coef in enumerate(self._estates[:, 0]):
+            ground_state += estate_coef * op.basis(len(self._evals), idx)
+        return ground_state
+
+    def calculate_frequencies(self, qubit_lo_freq=None, u_channel_lo=None):
```

Comment:
I think the name of this is good for this iteration. Something that might make sense next, as we refactor so that OPSystem holds a HamiltonianModel, is to break this function up: one (or several) in HamiltonianSystem that compute eigenvectors/eigenvalues/dressed energies for the drift Hamiltonian, and then another in OPSystem called something like `channel_frequencies`, which, if passed the argument `from_hamiltonian` will draw the relevant info from `HamiltonianModel`.

Reply:
Do you think this one should be made a private function for now? `def _calculate_frequencies` ?